src/main.rs

```diff
@@ -187,14 +213,130 @@ where command is one of:
-                        process::exit(1);
-                    });
+                let mut contract_analysis = type_check(&"transient", &mut ast, &mut db, false).unwrap_or_else(|e| {
+                    eprintln!("Type check error.\n{}", e);
+                    process::exit(1);
+                });
 
+                match argv.last() {
+                    Some(s) if s == "--output_analysis" => {
+                        println!("{}", contract_analysis.serialize());
+                    },
+                    _ => {}
+                }
+                
                 return
             },
+            "repl" => {
+                let mut db_conn = match ContractDatabaseConnection::memory() {
+                    Ok(db) => db,
+                    Err(error) => {
+                        eprintln!("Could not open vm-state: \n{}", error);
+                        process::exit(1);
+                    }
+                };
+
+                let mut outer_sp = db_conn.begin_save_point_raw();                    
+                let mut db = ContractDatabase::from_savepoint(outer_sp);
+
+                let mut vm_env = OwnedEnvironment::new(&mut db);
+                let mut exec_env = vm_env.get_exec_environment(None);
+
+                let mut analysis_db_conn = AnalysisDatabaseConnection::memory();
+
+                let mut reader = match linefeed::Interface::new("local-repl") {
+                    Ok(r) => r,
+                    Err(error) => panic!("Could not create linefeed: \n{}", error)
+                };
+
+                reader.set_report_signal(linefeed::Signal::Break, true);
+                reader.set_report_signal(linefeed::Signal::Continue, true);
+                reader.set_report_signal(linefeed::Signal::Interrupt, true);
+                reader.set_report_signal(linefeed::Signal::Suspend, true);
+                reader.set_report_signal(linefeed::Signal::Quit, true);
+                
+                match reader.set_prompt("> ") {
+                    Ok(r) => r,
+                    Err(error) => panic!("Could not create linefeed: \n{}", error)
+                };
+
+                loop {
+                    let content = match reader.read_line() {
+                        Ok(result) => match result {
+                            linefeed::ReadResult::Input(input) => input,
+                            linefeed::ReadResult::Signal(_) => process::exit(0),
+                            linefeed::ReadResult::Eof => process::exit(0),
+                        },
+                        Err(error) => panic!("Could not read line: \n{}", error)
+                    };
+                    
+                    if !content.trim().is_empty() {
+                        reader.add_history_unique(content.clone());
+                    }
+
+                    let mut ast = parse(&content).expect("Failed to parse program.");
+                    let mut analysis_db = analysis_db_conn.begin_save_point();
```

Comment:
Don't you have to commit your save points somewhere?  Maybe this is a question for @kantai

Reply:
yes, this would need to be committing these save points for the analysis database to update between invocations.

_however_ the way its currently written, no define statements would ever evaluate successfully, and any attempts to access a data map would also fail, so there's no data that would ever be written by the save point.

Reply:
Resolving this as REPL and eval_raw were moved into another branch. Statefulness will be an iterative feature later on.